src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1714 @@
+            ssm_state = discrete_A[:, :, :, i, :].transpose(0, 1) * ssm_state + deltaB_u[:, :, :, i, :].transpose(0, 1)  # (h b d state)
+            scan_output = torch.matmul(ssm_state.transpose(0, 1).to(dtype), C[:, :, i, :].unsqueeze(-1)) # (h b d 1) 
+            scan_outputs.append(scan_output[:, :, :, 0])
+        scan_output = torch.stack(scan_outputs, dim=-1)                                  # (h b d l)
+        scan_output = scan_output + (hidden_states * self.D[:, None, :, None])           # (h b d l)
+        scan_output = (scan_output * self.act(gate))
+
+        if use_cache:
+            cache_params.ssm_states[self.layer_idx] = ssm_state           
+
+        # 4. Final linear projection
+        contextualized_states = self.out_proj(scan_output.transpose(0, 1).reshape(batch_size, -1, seq_len).transpose(1, 2))             # (b l d)
+        return contextualized_states
+    
+
+    def forward(self, hidden_states, cache_params: HybridMambaAttentionDynamicCache = None):
+        if self.use_fast_kernels:
+            if not is_fast_path_available or "cuda" not in self.x_proj_weight.device.type:
+                raise ValueError(
+                    "Fast Mamba kernels are not available. Make sure to they are installed and that "
+                    "the mamba module is on a CUDA device. lease run 'pip install causal-conv1d>=1.2.0' "
+                    "and 'pip install mamba-ssm', or set use_fast_kernels=False in the model's config."
+                )
+            return self.cuda_kernels_forward(hidden_states, cache_params)
+        return self.slow_forward(hidden_states, cache_params)
+
+
+# Copied from transformers.models.mistral.modeling_mistral.MistralMLP with Mistral->Zamba
+class ZambaMLP(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.hidden_size = config.hidden_size
+        self.intermediate_size = config.intermediate_size
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
+        self.act_fn = ACT2FN[config.hidden_act]
+
+    def forward(self, x):
+        return self.down_proj(self.act_fn(self.up_proj(x)) * self.gate_proj(x))
+
+
+class ZambaAttentionDecoderLayer(nn.Module):
+    def __init__(self, config: ZambaConfig, layer_idx: Optional[int] = None):
+        super().__init__()
+        self.self_attn = ZAMBA_ATTENTION_CLASSES[config._attn_implementation](config, layer_idx)
+
+        self.feed_forward = ZambaMLP(config)
+        self.input_layernorm = ZambaRMSNorm(2 * config.hidden_size, eps=config.rms_norm_eps)
+        self.pre_ff_layernorm = ZambaRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        x_orig: torch.Tensor,
```

Comment:
Instead of passing it can you concatenate it before passing it to the decoder layer?

Reply:
It would be preferable to keep the concatenation inside the decoder layer. This layer is a modified version from the original Mistral implementation (as described in a previous reply above, there are a few modifications in terms of dimensions and attention temperature, compared to the original implementation in Mistral), so there is no particular gain in our view in moving the concatenation outside.

Reply:
Ok thanks for explaining, it won't hurt having it here as well, I was mainly concerned about breaking a bit our API consistency when adding new arguments in the forward pass of a module. In this case I think that's fine but let's wait for a core maintainers review - Can we also change the name to `original_hidden_states`?

Reply:
The name of this variable was updated to `original_hidden_states`.

Please re-open this comment chain if you have further comments!